File: cpp/clion/Prog_미로탈출명령어.cpp
#include <iostream>
#include <string>
#include <vector>
using namespace std;

int R,C,sr,sc,er,ec,K;
string ans = "";
int dr[] = {1, 0, 0, -1};
int dc[] = {0, -1, 1, 0};
string m = "dlru";
string tmp = "";

bool inRange(int r, int c) {
    return r > 0 and r <= R and c > 0 and c <= C;
}

bool dfs(int r, int c) {
    if (tmp.size() == K) {
        if (r == er && c == ec) {
            ans = tmp;
            return true;
        }
        return false;
    }
    for (int i=0;i<4;i++) {
        int nr = r + dr[i];
        int nc = c + dc[i];
        if (!inRange(nr,nc)) continue;
        tmp.push_back(m[i]);
        if (dfs(nr, nc)) return true;
        tmp.pop_back();
    }
    return false;
}

string solution(int n, int m, int x, int y, int r, int c, int k) {
    R = n; C = m; sr = x; sc = y; er = r; ec = c; K = k;
    if (dfs(sr, sc)) {
        return ans;
    }
    return "impossible";
}

int main() {
    cout << solution(3, 4, 2, 3, 3, 1, 5);
}

File: python/Prog_미로탈출명령어.py
import sys
from collections import defaultdict
sys.setrecursionlimit(3000)

R = 0
C = 0
sr = 0
sc = 0
er = 0
ec = 0
K = 0
ans = ""
visited = [[]]


dr = [1,0,0,-1]
dc = [0,-1,1,0]
m = "dlru"

def in_range(r,c) :
    return 0<r<=R and 0<c<=C

def dfs(r,c, tmp) :
    global ans
    if visited[r][c][len(tmp)] :
        return False
    visited[r][c][len(tmp)] = True
    
    if len(tmp) == K :
        if r == er and c == ec : 
            ans = tmp
            return True
        return False
    
    for i in range(4) :
        nr = r + dr[i]
        nc = c + dc[i]
        if not in_range(nr,nc) : continue
        if dfs(nr,nc, tmp+m[i]) :
            return True
        
    return False
            

def solution(n, m, x, y, r, c, k):
    global R,C, sr,sc, er,ec, K, visited
    R = n
    C = m
    sr = x
    sc = y
    er = r
    ec = c
    K = k
    visited = [ [defaultdict(bool) for _ in range(C+1)] for _ in range(R+1) ]
    
    if dfs(sr,sc, "") :
        return ans
    return "impossible"

print(solution(3,3,0,0,2,0,3))